src/codegen.cpp

```diff
@@ -2718,18 +2718,24 @@ static bool emit_builtin_call(jl_codectx_t &ctx, jl_cgval_t *ret, jl_value_t *f,
             auto len = emit_arraylen(ctx, obj, ary_ex);
             jl_value_t *ety = jl_tparam0(sty);
             Value *elsize;
+            size_t elsz = 0, al = 0;
+            bool isboxed = !jl_islayout_inline(ety, &elsz, &al);
             if (!jl_has_free_typevars(ety)) {
-                if (!jl_array_store_unboxed(ety)) {
+                if (isboxed) {
                     elsize = ConstantInt::get(T_size, sizeof(void*));
                 }
                 else {
-                    elsize = ConstantInt::get(T_size, jl_datatype_size(ety));
+                    elsize = ConstantInt::get(T_size, elsz);
```

Comment:
We should have a more convenient `jl_type_size` function for just getting the size.